### ISSUE TITLE
Getting 'bluepy.btle.BTLEDisconnectError: Device disconnected' exception after weighing

### ISSUE THREAD

**Issue by risa2000** [lolouk44/xiaomi_mi_scale#12]
I am running the script on RPi3b (no+) with _Xiaomi Mi Body Composition Scale 2_ (`XMTZC05HM`) simply as a background process by:
```sh
$ sudo python3 Xiaomi_Scale.py
```
The script starts ok, but after I weigh myself and the data are published to MQTT server, after a while the script crashes with
```python
Traceback (most recent call last):
  File "Xiaomi_Scale.py", line 180, in <module>
    main()
  File "Xiaomi_Scale.py", line 173, in main
    scanner.scan(5)
  File "/usr/local/lib/python3.7/dist-packages/bluepy/btle.py", line 854, in scan
    self.stop()
  File "/usr/local/lib/python3.7/dist-packages/bluepy/btle.py", line 803, in stop
    self._mgmtCmd(self._cmd()+"end")
  File "/usr/local/lib/python3.7/dist-packages/bluepy/btle.py", line 309, in _mgmtCmd
    rsp = self._waitResp('mgmt')
  File "/usr/local/lib/python3.7/dist-packages/bluepy/btle.py", line 362, in _waitResp
    raise BTLEDisconnectError("Device disconnected", resp)
bluepy.btle.BTLEDisconnectError: Device disconnected
```

**Comment by lolouk44** [comment]
Thanks. A few users have reported issues when using a Raspberry Pi and said that it works fine when using a Bluetooth dongle.
I personally don't use one for the script so can't test, and in any case the issue appears to lie with bluepy which I don't maintain.
I'm working on another Bluetooth project at the moment, I'll review this later to see if I can change the way I connect to the scale in case there's something in there.

**Comment by risa2000** [comment]
Just for the record, I disabled the onboard BT chip and connected BT LE 4.0 dongle:
```
ID 050d:065a Belkin Components F8T065BF Mini Bluetooth 4.0 Adapter
```
and I am getting the same error, after a while.
```python
Traceback (most recent call last):
  File "Xiaomi_Scale.py", line 180, in <module>
    main()
  File "Xiaomi_Scale.py", line 173, in main
    scanner.scan(5)
  File "/usr/local/lib/python3.7/dist-packages/bluepy/btle.py", line 854, in scan
    self.stop()
  File "/usr/local/lib/python3.7/dist-packages/bluepy/btle.py", line 803, in stop
    self._mgmtCmd(self._cmd()+"end")
  File "/usr/local/lib/python3.7/dist-packages/bluepy/btle.py", line 309, in _mgmtCmd
    rsp = self._waitResp('mgmt')
  File "/usr/local/lib/python3.7/dist-packages/bluepy/btle.py", line 362, in _waitResp
    raise BTLEDisconnectError("Device disconnected", resp)
bluepy.btle.BTLEDisconnectError: Device disconnected
```
It looks as if the BT stack remains connected to the scale, and once it switches into "stand-by", it disconnects and triggers this error.

**Comment by benleb** [comment]
Sorry, just because I am curious (and also own this scale and develop code to communicate with it)... is this issue just to inform or what’s the actual problem? As far as I understand everything is done until the error occurs and as the script runs independently every time, the only issue is the error message?

**Comment by risa2000** [comment]
@benleb if the question was to me, I expect it to be an issue because it crashes the script, which otherwise, as it seems, should be running indefinitely, and because it leaves the BT stack and network stack in some unstable state (sometimes the script does not connect to MQTT after the crash or to the scale).

**Comment by benleb** [comment]
@risa2000 yes, thanks! If it should run indefinitely I can see the problem :D If you are not afraid of some alpha-state code, you can try [miblepy](https://github.com/benleb/miblepy) which uses the logic/bytes from @lolouk44 (again, thanks for your work!) but runs every few minutes as a service instead of a continuously running daemon, avoiding problems like this here.

**Comment by lolouk44** [comment]
yes I see the issue as well, I've commented the `try` function to keep it running eternally.
Replace this
https://github.com/lolouk44/xiaomi_mi_scale/blob/422b6ed9bdc18ac518abde9861ad91b5c56e401d/src/Xiaomi_Scale.py#L171-L177

with
```
	while True:
		try:
			scanner.scan(5)
		except:
			sys.stderr.write("Error while running the script, continuing. If you see this message too often/constantly there is probably a real issue...\n")
			pass
		time.sleep(30)
```
I've changed the sleep to 30 sec so the script does not lock the BT dongle, feel free to change that value as per your preferences

**Comment by lolouk44** [comment]
just pushed an update to my repo with a minor change regarding the BT connection,using `passive=True` in `scanner.scan` after seeing somewhere that it may help with Raspberry Pis.
@risa2000 can you test and report pls?

**Comment by risa2000** [comment]
@lolouk44 
It seems you actually comitted a typo (syntax error) with the `TIME_INTERVAL` addition. There is also a problem with your handling of MQTT connection. You setup an asynchronous handler `_on_connect`, but you do not wait for the connection and check the flag `self.connected` asynchronously as well. Which means that in my case MQTT is always late and the program aborts with
```python
                if not self.connected:
                        sys.stderr.write('Not connected to MQTT server\n')
```
I could hardcode a workaround, but this architecture probably needs some redesign.

The last thing which I found strange is using the hard-tabs in the code. I am not sure if you are really using 8 spaces for one indent in Python, or your just redefined hard-tabs locally to 4 spaces, but it is quite unusual (in Python ecosystem).

Anyway, after the changes I made, I am still seeing the disconnect exception (if not handled by try-except). I guess you can "mask" it by adding try-except around it and I wonder, is raising
```
BTLEDisconnectError("Device disconnected", resp)
```
the standard way of handling the device disconnection from BTLE stack?

**Comment by risa2000** [comment]
@benleb Thanks for the tip for your repo as I happen to own also `LYWSDCGQ` hygrothermo sensors (currently integrated into homeassitant by a custom sensor script), I will definitely check it out.

**Comment by lolouk44** [comment]
> It seems you actually comitted a typo (syntax error) with the `TIME_INTERVAL` addition

Yes I did but fixed it later (I actually committed the fix but didn't originale pushed it)

> The last thing which I found strange is using the hard-tabs in the code. I am not sure if you are really using 8 spaces for one indent in Python, or your just redefined hard-tabs locally to 4 spaces, but it is quite unusual (in Python ecosystem).

Yes my code originally had tabs instead of spaces and I kept it that way. Should probably amend indeed 


> Anyway, after the changes I made, I am still seeing the disconnect exception (if not handled by try-except). I guess you can "mask" it by adding try-except around it and I wonder, is raising
> 
> ```
> BTLEDisconnectError("Device disconnected", resp)
> ```
> 
> the standard way of handling the device disconnection from BTLE stack?

I have to be honest, this is my first foray into the Bluetooth world. I could not find find a solution online to connect to the scale so I took pieces of code I found online and put them together. There is definitely a better way of doing this. Already the with delegate bit is deprecated from bluepy... Feel free to fix my code and submit a PR 😊

**Comment by lolouk44** [comment]
Not too sure what's going on with the Raspberry Pi. I used one for testing and I do get lots of issues.
I've made changes to better catch errors and added some bluetooth interface reset
Can you try 0.1.1 and let me know how it goes?

**Comment by lolouk44** [comment]
I'll close this as it appears my last changes under 0.1.1 do improve connection reliability a lot. If issues still occur, they are most likely an issue with bluepy...

**Comment by risa2000** [comment]
Just for the record, I tried the latest version, but it does not work for me. The repeated `btle disconnected` messages are the masked exceptions. Then when I weighed myself, the message related to this is `Could not publish to MQTT, Disconnecting...`.

BT data were received, but lost in the handler, because the MQTT was not connected. As I wrote above, you might want to rethink how you handle the connections to MQTT and to BT and possibly handle them separately, so you do not have to tear down MQTT when BT fails, and vice versa. Plus let the script actually connect to MQTT, before tearing it down.
```
btle disconnected: Device disconnected
btle disconnected: Device disconnected
btle disconnected: Device disconnected
btle disconnected: Device disconnected
btle disconnected: Device disconnected

Could not publish to MQTT, Disconnecting...
Error while running the script, continuing...
MQTT connection: Connection Accepted.
btle disconnected: Device disconnected
...
```
